platforms/jvm/code-quality/src/main/groovy/org/gradle/api/plugins/quality/Pmd.java

```diff
@@ -81,6 +79,9 @@ public Pmd() {
         this.incrementalAnalysis = objects.property(Boolean.class);
         this.maxFailures = objects.property(Integer.class);
         this.threads = objects.property(Integer.class);
+
+        DirectoryProperty temporaryDir = objects.directoryProperty().fileProvider(getProject().provider(this::getTemporaryDir));
```

Comment:
I haven't thought about this deeply, but I wonder if we should...
1. Migrate `getTemporaryDir` to a `DirectoryProperty`
2. Or add a `getTemporaryDirectory` that's a `DirectoryProperty`

And then `incrementalCacheFile` could be a `RegularFileProperty` and wired to that.

Reply:
That would make sense. From implementation point of view adding `getTemporaryDirectory()` would be easier. 
I guess it should be `Provider<Directory` since users should not set it, or at least it should be finalized.

Due to complexity I think we should do that as a separate PR. I've opened https://github.com/gradle/gradle/issues/30082 and added 9.0-RC1 milestone so we don't forget about it.